source/loader/gamecard.cpp

```diff
@@ -249,58 +271,141 @@ class Adaptor3DSXToNCCH : public HLE::PXI::FS::File {
             throw std::runtime_error("Can't create/truncate game cards");
         }
 
-        auto ret = file->Open(context, flags);
-        if (ret != std::tie(HLE::OS::RESULT_OK)) {
-            return ret;
+        auto open_ret = file->Open(context, flags);
+        if (open_ret != std::tie(HLE::OS::RESULT_OK)) {
+            return open_ret;
         }
 
-        return ret;
-    }
-
-    FileFormat::Dot3DSX::Header Read3DSXHeader(HLE::PXI::FS::FileContext& context) {
-        // TODO: Use serialization interface instead
-        FileFormat::Dot3DSX::Header dot3dsxheader;
         auto ret = file->Read(context, 0, sizeof(dot3dsxheader), HLE::PXI::FS::FileBufferInHostMemory(dot3dsxheader));
-        if (ret != HLE::OS::OS::ResultAnd<uint32_t> { HLE::OS::RESULT_OK, sizeof(dot3dsxheader) }) {
-            throw std::runtime_error("Failed to read 3DSX header");
+        if (ret != std::tuple { HLE::OS::RESULT_OK, (unsigned int)sizeof(dot3dsxheader) }) {
+            throw Mikage::Exceptions::Invalid("Could not read 3DSX header");
         }
 
-        // TODO: Check magic, check header size
+        // Read and validate header(s), populate SMDH data
+        if (dot3dsxheader.header_size == sizeof(FileFormat::Dot3DSX::Header)) {
+            icon_data.magic = { 'S', 'M', 'D', 'H' };
+            constexpr std::array<char16_t, 9> name { u"Homebrew" };
 
-        return dot3dsxheader;
-    }
+            for (auto& app_title : icon_data.application_titles) {
+                ranges::copy(name, app_title.long_description.begin());
+                ranges::copy(name, app_title.short_description.begin());
+                ranges::copy(name, app_title.publisher_name.begin());
+            }
+
+            using SMDHFlag = FileFormat::SMDH::SMDHFlags;
+
+            icon_data.application_settings.region_flags = 0xFFFFFFFF; // no region lock
+            icon_data.application_settings.flags = Meta::to_underlying(SMDHFlag::Visible | SMDHFlag::DisableSaveBackup);
+            // Give this SMDH a placeholder icon
+            ranges::copy(GameCardConstants::MikageIconSmall, icon_data.small_icon.begin());
+            ranges::copy(GameCardConstants::MikageIconLarge, icon_data.large_icon.begin());
+        } else if (dot3dsxheader.header_size == sizeof(FileFormat::Dot3DSX::Header) + sizeof(FileFormat::Dot3DSX::SecondaryHeader)) {
+            dot3dsx_secondary_header = FileFormat::Dot3DSX::SecondaryHeader{};
+            auto ret = file->Read(context, sizeof(FileFormat::Dot3DSX::Header), sizeof(FileFormat::Dot3DSX::SecondaryHeader), HLE::PXI::FS::FileBufferInHostMemory(dot3dsx_secondary_header));
```

Comment:
There's a pretty bad bug hiding here that I'll fix separately: `HLE::PXI::FS::FileBufferInHostMemory` will write over the `std::optional`, not the actual header data. I'm surprised this is working at all, but perhaps std::optional just happens to store the contained object first (so we aren't overwriting the internal state by pure luck).

`FileBufferInHostMemory` should be more constrained to make this fail to compile, so it's good that we caught it this way. I'll add the fix in a bit.